File: preparation/src/shared/infrastructure/events/RabbitmqDomainEventsConsumer.cs
/* 
 * 
 * Libraries
 *
*/

using System.Text;
using preparation.src.shared.domain;
using RabbitMQ.Client;
using RabbitMQ.Client.Events;

/* 
 * 
 * Class
 *
*/

namespace preparation.src.shared.infrastructure {

    public class RabbitmqDomainEventsConsumer { 

        /* 
         * 
         * Constants
         *
        */

        private const string RETRY_COUNT_FIELD_NAME = "retryCount";

        /* 
         * 
         * Attributes
         *
        */

        private RabbitmqEventBusConnector     eventBusConnector;
        private DomainEventsInformation       eventsInformation;
        private RabbitmqExchangeNameFormatter exchangeNameFormatter;
        private JsonDomainEventDeserializer   domainEventDeserializer;
        private int                           messageDeliveryMode;
        private int                           maximumNumberOfRetries;
        private IServiceProvider              serviceProvider;

        /* 
         * 
         * Methods
         *
        */

        public RabbitmqDomainEventsConsumer( 
            RabbitmqEventBusConnector     eventBusConnector,
            DomainEventsInformation       eventsInformation,
            RabbitmqExchangeNameFormatter exchangeNameFormatter,
            JsonDomainEventDeserializer   domainEventDeserializer,
            IServiceProvider              serviceProvider 
        ) {
            this.eventBusConnector       = eventBusConnector;
            this.eventsInformation       = eventsInformation;
            this.exchangeNameFormatter   = exchangeNameFormatter;
            this.domainEventDeserializer = domainEventDeserializer;
            this.serviceProvider         = serviceProvider;
            messageDeliveryMode = Int32.Parse( Environment.GetEnvironmentVariable( 
                "EVENT_BUS_MESSAGE_DELIVERY_MODE"
            ) );
            maximumNumberOfRetries = Int32.Parse( Environment.GetEnvironmentVariable( 
                "MAXIMUM_NUMBER_OF_EVENT_BUS_RETRIES"
            ) );
        }

        public async Task consume() {
            foreach( DomainEventInformation eventInformation in eventsInformation.getAll() ) {
                if( eventInformation.isConsumedEvent() ) {
                    await Task.Run( () => {
                        consumeEvent( eventInformation );
                    } );
                }
            }
        }

        private void consumeEvent( DomainEventInformation eventInformation ) {
            // Variables
            IModel                    channel;
            EventingBasicConsumer     consumer;
            string                    queueName;
            DomainEvent               domainEvent;
            string                    body;
            byte[]                    message;
            DomainEventSubscriberBase eventSubscriber;
            IServiceScope             scope;
            // code
            channel   = eventBusConnector.getChannel();
            consumer  = new EventingBasicConsumer( channel );
            queueName = eventInformation.formatRabbitmqQueueName();
            scope     = serviceProvider.CreateScope();
            consumer.Received += async ( model, deliverEventArgs ) => {
                message     = deliverEventArgs.Body.ToArray();
                body        = Encoding.UTF8.GetString( message );
                domainEvent = domainEventDeserializer.deserialize( body );
                try {
                    foreach( Type subscriberClass in eventInformation.getSubscriberClasses() ) {
                        eventSubscriber = ( DomainEventSubscriberBase )scope.ServiceProvider.GetRequiredService( subscriberClass );
                        await eventSubscriber.handleEvent( domainEvent );
                    }
                } catch( DomainException ) {
                    handleConsumptionError( deliverEventArgs, domainEvent );
                }
                channel.BasicAck( deliverEventArgs.DeliveryTag, false );
            };
            channel.BasicConsume(
                queue    : queueName,
                autoAck  : false,
                consumer : consumer
            );
        }

        private void handleConsumptionError( 
            BasicDeliverEventArgs deliverEventArgs, 
            DomainEvent           domainEvent 
        ) {
            // Variables
            string eventName;
            // code
            eventName = domainEvent.getEventName();
            if( hasBeenRetriedTooMuch( deliverEventArgs ) ) {
                sentToDeatLetter( deliverEventArgs, eventName );
            } else {
                sendToRetry( deliverEventArgs, eventName );
            }
        }

        private void sendToRetry( 
            BasicDeliverEventArgs deliverEventArgs, 
            string                eventName
        ) {
            // Variables
            int              retryCount;
            string           messageBody;
            IBasicProperties messageProperties;
            IModel           channel;
            byte[]           message;
            // Code
            retryCount = getRetryCount( deliverEventArgs );
            channel    = eventBusConnector.getChannel();
            messageProperties = channel.CreateBasicProperties();
            messageProperties.DeliveryMode = ( byte )messageDeliveryMode;
            messageProperties.Headers = new Dictionary<string, object>() { {
                RETRY_COUNT_FIELD_NAME, ( retryCount + 1 )
            } };
            message     = deliverEventArgs.Body.ToArray();
            messageBody = Encoding.UTF8.GetString( message );
            sendMessage( eventName, messageBody, messageProperties );
            channel.Close();
        }

        private bool hasBeenRetriedTooMuch( BasicDeliverEventArgs deliverEventArgs ) {
            // Variables
            int retryCount;
            // Code
            retryCount = getRetryCount( deliverEventArgs );
            return retryCount >= ( maximumNumberOfRetries - 1 );
        }

        private int getRetryCount( BasicDeliverEventArgs deliverEventArgs ) {
            // Variables
            int                         retryCount;
            IDictionary<string, object> messageHeaders;
            // Code
            messageHeaders = deliverEventArgs.BasicProperties.Headers;
            retryCount     = 0;
            if( messageHeaders != null && 
                messageHeaders.ContainsKey( RETRY_COUNT_FIELD_NAME ) 
            ) {
                retryCount = ( int )messageHeaders[RETRY_COUNT_FIELD_NAME];
            }
            return retryCount;
        }

        private void sentToDeatLetter( 
            BasicDeliverEventArgs deliverEventArgs,
            string                eventName
        ) {
            // Variables
            string exchangeName;
            string messageBody;
            byte[] message;
            // Code
            message      = deliverEventArgs.Body.ToArray();
            messageBody  = Encoding.UTF8.GetString( message );
            exchangeName = exchangeNameFormatter.formatDeadLetter( eventName );
            sendMessage( exchangeName, messageBody );
        }

        private void sendMessage( 
            string            exchangeName, 
            string            body,
            IBasicProperties? properties = null 
        ) {
            // Variables
            IModel channel;
            // code
            channel = eventBusConnector.getChannel();
            if( properties == null ) {
                properties = channel.CreateBasicProperties();
                properties.DeliveryMode = ( byte )messageDeliveryMode;
            }
            channel.BasicPublish(
                exchange        : exchangeName,
                routingKey      : "", 
                body            : UTF8Encoding.UTF8.GetBytes( body ), 
                basicProperties : properties
            );
            channel.Close();
        }
    
    }

}

File: backoffice/src/shared/infrastructure/events/RabbitmqDomainEventsConsumer.py
"""
 *
 * Libraries 
 *
"""

import os
from typing                                   import Any
from src.shared.domain                        import DomainEvent
from src.shared.domain                        import DomainEventSubscriber
from src.shared.domain                        import DomainException
from .DomainEventsInformation                 import DomainEventsInformation
from injector                                 import inject
from threading                                import Thread
from amqpstorm                                import Message
from amqpstorm                                import Channel
from ..eventBus.RabbitmqEventBusConnector     import RabbitmqEventBusConnector
from .JsonDomainEventDeserializer             import JsonDomainEventDeserializer
from .DomainEventInformation                  import DomainEventInformation
from ..eventBus.RabbitmqExchangeNameFormatter import RabbitmqExchangeNameFormatter
from injector                                 import Injector

"""
 *
 * Class
 *
"""

class RabbitmqDomainEventsConsumer:

    """
     *
     * Constants 
     *
    """

    __RETRY_COUNT_FIELD_NAME = 'retryCount'

    """
     *
     * Attributes 
     *
    """

    __eventBusConnector       : RabbitmqEventBusConnector
    __eventsInformation       : DomainEventsInformation
    __domainEventDeserializer : JsonDomainEventDeserializer
    __exchangeNameFormatter   : RabbitmqExchangeNameFormatter
    __messageDeliveryMode     : int
    __maximumNumberOfRetries  : int
    __injector                : Injector

    """
     *
     * Methods 
     *
    """

    @inject
    def __init__( 
        self, 
        eventBusConnector       : RabbitmqEventBusConnector,
        eventsInformation       : DomainEventsInformation,
        domainEventDeserializer : JsonDomainEventDeserializer,
        exchangeNameFormatter   : RabbitmqExchangeNameFormatter,
        injector                : Injector
    ) -> object:
        self.__eventBusConnector       = eventBusConnector
        self.__eventsInformation       = eventsInformation
        self.__domainEventDeserializer = domainEventDeserializer
        self.__exchangeNameFormatter   = exchangeNameFormatter
        self.__messageDeliveryMode     = int( os.getenv( 'EVENT_BUS_MESSAGE_DELIVERY_MODE' ) )
        self.__maximumNumberOfRetries  = int( os.getenv( 'MAXIMUM_NUMBER_OF_EVENT_BUS_RETRIES' ) )
        self.__injector                = injector

    def consume( self ) -> None:
        for eventInformation in self.__eventsInformation.getAll():  
            if eventInformation.isConsumedEvent():          
                Thread( target = self.__consumeEvent, args = ( eventInformation, ) ).start()

    def __consumeEvent( self, eventInformation : DomainEventInformation ) -> None:
        # Variables
        channel   : Channel
        queueName : str
        # Code
        channel   = self.__eventBusConnector.getChannel()
        queueName = eventInformation.formatRabbitmqQueueName()
        channel.basic.consume( 
            queue    = queueName,
            callback = lambda message:
                self.__callback( message, eventInformation )
        )
        channel.start_consuming()
    
    def __callback( 
        self, 
        message          : Message, 
        eventInformation : DomainEventInformation 
    ) -> None:
        # Variables
        domainEvent     : DomainEvent
        eventSubscriber : DomainEventSubscriber
        # Code
        domainEvent = self.__domainEventDeserializer.deserialize( message.body )
        try:
            for subscriberClass in eventInformation.getSubscriberClasses():
                eventSubscriber = self.__injector.get( subscriberClass )
                eventSubscriber.handleEvent( domainEvent )
        except DomainException:
            self.__handleConsumptionError( message, domainEvent )
        message.ack()
    
    def __handleConsumptionError( self, message : Message, domainEvent : DomainEvent ) -> None:
        # Variables
        eventName : str
        # Code
        eventName = domainEvent.getEventName()
        if self.__hasBeenRetriedTooMuch( message ):
            self.__sentToDeatLetter( message, eventName )
        else:
            self.__sendToRetry( message, eventName )

    def __hasBeenRetriedTooMuch( self, message : Message ) -> bool:
        # Variables
        retryCount : int
        # Code
        retryCount = self.__getRetryCount( message )
        return retryCount >= ( self.__maximumNumberOfRetries - 1 )
    
    def __getRetryCount( self, message : Message ) -> int:
        # Variables
        messageHeaders : dict[str, str | int]
        retryCount     : int
        # Code
        messageHeaders = message.properties['headers']
        retryCount     = 0
        if messageHeaders:
            retryCount = messageHeaders[self.__RETRY_COUNT_FIELD_NAME]
        return retryCount

    def __sendToRetry( self, message : Message, eventName : str ) -> None:
        # Variables
        retryCount        : int
        messageBody       : str
        messageProperties : dict[str, Any]
        # Code
        retryCount = self.__getRetryCount( message )
        messageProperties = {
            'delivery_mode' : self.__messageDeliveryMode,
            'headers'       : { self.__RETRY_COUNT_FIELD_NAME : ( retryCount + 1 ) }
        }
        messageBody = message.body
        self.__sendMessage( eventName, messageBody, messageProperties )
    
    def __sentToDeatLetter( self, message : Message, eventName : str ) -> None:
        # Variables
        exchangeName : str
        messageBody  : str
        # Code
        messageBody  = message.body
        exchangeName = self.__exchangeNameFormatter.formatDeadLetter( eventName )
        self.__sendMessage( exchangeName, messageBody )

    def __sendMessage( 
        self,
        exchangeName : str,
        body         : str,
        properties   : dict[str, Any] = {}
    ) -> None:
        # Variables
        channel : Channel    
        # Code                
        if not properties:
            properties = { 'delivery_mode' : self.__messageDeliveryMode }
        channel = self.__eventBusConnector.getChannel()        
        channel.basic.publish( 
            exchange    = exchangeName,
            routing_key = '',
            body        = body,
            properties  = properties
        )
        channel.close()